ngraph/core/src/op/interpolate.cpp

```diff
@@ -111,143 +114,538 @@ namespace ngraph
+        }
+
+        std::vector<T> result;
+
+        if (pad_len > rank)
+        {
+            result.insert(result.end(), p.begin(), p.begin() + rank);
+        }
+        else
+        {
+            result = p;
+            result.insert(result.end(), rank - pad_len, T{});
+        }
+
+        return result;
+    }
+
+    struct InfoToCallReference
+    {
+        Shape input_shape;
+        Shape padded_input_shape;
+        Shape out_shape;
+        std::vector<size_t> pads_begin;
+        std::vector<size_t> pads_end;
+        std::vector<int64_t> axes;
+        std::vector<float> scales;
+    };
+
+    using InterpolateV4Attrs = op::v4::Interpolate::InterpolateAttrs;
+
+    InfoToCallReference get_info_to_call_reference(const HostTensorVector& args,
+                                                   const InterpolateV4Attrs& attrs)
+    {
+        using ShapeCalcMode = ngraph::op::v4::Interpolate::ShapeCalcMode;
+
+        Shape input_shape{args[data_port]->get_shape()};
+        std::size_t input_rank = input_shape.size();
+
+        auto axes = get_axes_vector(args);
+        std::size_t num_of_axes = axes.size();
+
+        auto scales = get_scales_vector(args, input_shape, attrs, axes);
+
+        auto pads_begin = correct_pad(attrs.pads_begin, input_rank);
+        auto pads_end = correct_pad(attrs.pads_end, input_rank);
+
+        auto padded_input_shape_vector = get_padded_input_shape(input_shape, pads_begin, pads_end);
+        std::vector<std::size_t> out_shape_vector;
+
+        if (attrs.shape_calculation_mode == ShapeCalcMode::scales)
+        {
+            out_shape_vector = shape_infer_with_scales(padded_input_shape_vector, axes, scales);
+        }
+        else
+        {
+            auto target_shape = get_target_shape_vector(args, num_of_axes);
+            out_shape_vector =
+                shape_infer_with_target_shape(padded_input_shape_vector, axes, target_shape);
+        }
+
+        Shape out_shape{out_shape_vector};
+        Shape padded_input_shape{padded_input_shape_vector};
+
+        return {input_shape, padded_input_shape, out_shape, pads_begin, pads_end, axes, scales};
+    }
+
+    template <element::Type_t ET>
+    inline bool evaluate(const HostTensorVector& args,
+                         const HostTensorPtr& out,
+                         const op::v4::Interpolate::InterpolateAttrs& attrs)
+    {
+        using T = typename element_type_traits<ET>::value_type;
+
+        auto info_for_reference = get_info_to_call_reference(args, attrs);
+
+        out->set_element_type(args[0]->get_element_type());
+        out->set_shape(info_for_reference.out_shape);
+
+        std::vector<T> padded_input_data(shape_size(info_for_reference.padded_input_shape), T{});
+
+        CoordinateTransform input_transform(info_for_reference.input_shape);
+        CoordinateTransform padded_transform(info_for_reference.padded_input_shape);
+
+        const T* data_ptr = args[0]->get_data_ptr<ET>();
```

Comment:
You can remove template arguments for this method if you avoid this type specific code.

In this method you can use `memcpy + element_size + pointer to byte` instead of `padded_data_ptr[padded_transform.index(padded_coord)] = data_ptr[input_transform.index(input_coord)];`  in order to avoid data specific types.

Reply:
But the function evaluate() should have a template argument, because `linear` and `cubic` mode use element type to calculate inner sum.

Reply:
`linear` and `cubic` modes are used inside  `runtime::reference::interpolate<T>(...)` method.

Do you mean that this method: `runtime::reference::interpolate<T>(...)` should have a template argument?

Reply:
We cannot remove template argument of evaluate, because this template argument is used in `evaluate_interpolate_v4` switch case, and we cannot remove this switch case, because Interpolate-4 specification specifies `fp16`, `fp32`, `int8` as admissible types of elements of an input tensor. 

But we can reduce a body of the template function 
`inline bool evaluate(const HostTensorVector& args, const HostTensorPtr& out,  const op::v4::Interpolate::InterpolateAttrs& attrs)`

Reply:
`validate_and_infer_types` (not evaluate method) should check available arguments.

`evaluate` method can work only on data types which were validated by the `validate_and_infer` method.

**It means that switch case and template arguments are redundant!**

I don't want to reduce the body of a template `evaluate` method, I want to remove template arguments for `evaluate` method. Template arguments are not required for this method.
You can add assert if you want to recheck supported data_types, but in any case we don't need template arguments and template code. `HostTensor` contains all information which is needed for it.

Reply:
Maybe switch case is redundant in the function `evaluate_interpolate_v4`, but **then similar switch case is needed in `bool op::v4::Interpolate::evaluate`, because we should call various instances of `interpolate<T>` for various types `T` of elements of an input tensor.**

Reply:
But you don't need template argument for this evaluate method. You need to get template argument only for `interpolate<T>` call. It is different things.
Moreover you can extend interpolate method and add an argument with element type. It allow to move a switch case to details of implementation.

Reply:
Now bodies of the template inline function `evaluate` and the body of `evaluate_interpolate_v4` are moved into the `Inerpolate::evaluate` method.